osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineBlueprintContainer.cs

```diff
@@ -163,14 +164,14 @@ private void updateStacking()
 
         protected override SelectionHandler<HitObject> CreateSelectionHandler() => new TimelineSelectionHandler();
 
-        protected override SelectionBlueprint<HitObject> CreateBlueprintFor(HitObject item)
+        protected sealed override SelectionBlueprint<HitObject> CreateBlueprintFor(HitObject item)
         {
-            return new TimelineHitObjectBlueprint(item)
-            {
-                OnDragHandled = e => hitObjectDragged = e != null,
-            };
+            return CreateTimelineBlueprintFor(item)?.With(b => b.OnDragHandled = e => hitObjectDragged = e != null);
         }
 
+        [CanBeNull]
+        public virtual TimelineHitObjectBlueprint CreateTimelineBlueprintFor(HitObject hitObject) => new TimelineHitObjectBlueprint(hitObject);
```

Comment:
If we're going to go and say that it's fine for the timeline to look differently on different rulesets, then I can already tell you that this API is too constrained and we'll need a higher-level primitive anyway to replace the _entirety_ of the timeline.

Why, you ask? One of the most frequent pieces of feedback on scrolling rulesets (taiko and mania) are that scroll speed changes should be visible on the timeline. And osu!'s displays above hitobjects do not make sense because _that's not how speed changes work on those rulesets_ - they affect the _entire playfield_. So they should probably look closer to the red timing change markers.

So I think this sort of thing is too short-sighted and we'll have to tear it out sooner than later anyway. Not opposed to the idea, yet wholly opposed to this execution.

Reply:
This is valuable feedback.

Then we'll need to expand the API to allow showing additional things on the timeline apart from hit objects. I'm thinking such taiko/mania scroll speed changes would probably have to sit next to the `TimelineTimingChangeDisplay` in the drawable tree.

My question is to what degree do we want rulesets to replace the timeline? You said "replace the *entirety* of the timeline", but maybe thats overdoing it? I think all rulesets will still share a lot of elements of the timeline like the zoom behaviour, centre marker, timing change display, waveform, ticks, etc. Whatever implementation we go with, it should be easy to add these elements to your ruleset's timeline, so we don't end up with lots of code duplication.

Reply:
> Whatever implementation we go with, it should be easy to add these elements to your ruleset's timeline, so we don't end up with lots of code duplication.

I mean obviously, yes - that said, composition over inheritance please. I don't want to see any `BaseTimeline` shaped monstrosities. Just have the required components readily available for reuse.